src/graphql/operations/spaces.ts

```diff
@@ -1,47 +1,55 @@
-import db from '../../helpers/mysql';
-import { formatSpace } from '../helpers';
+import graphqlFields from 'graphql-fields';
+import { fetchSpaces } from '../helpers';
 
-export default async function(_parent, args) {
-  const { where = {} } = args;
-  let queryStr = '';
-  const params: any[] = [];
+export default async function(_parent, args, _context, info) {
+  try {
+    // maximum sum of parents and children to fetch
+    // only used if parent or children fields other than id are requested
+    const maxSubRequests = 100;
+    const requestedFields = info ? graphqlFields(info) : {};
 
-  const fields = ['id'];
-  fields.forEach(field => {
-    if (where[field]) {
-      queryStr += `AND s.${field} = ? `;
-      params.push(where[field]);
-    }
-    const fieldIn = where[`${field}_in`] || [];
-    if (fieldIn.length > 0) {
-      queryStr += `AND s.${field} IN (?) `;
-      params.push(fieldIn);
-    }
-  });
+    const spaces = await fetchSpaces(args);
 
-  let orderBy = args.orderBy || 'created_at';
-  let orderDirection = args.orderDirection || 'desc';
-  if (!['created_at', 'updated_at', 'id'].includes(orderBy))
-    orderBy = 'created_at';
-  orderDirection = orderDirection.toUpperCase();
-  if (!['ASC', 'DESC'].includes(orderDirection)) orderDirection = 'DESC';
+    if (
+      (requestedFields.parent &&
+        Object.keys(requestedFields.parent).some(key => key !== 'id')) ||
+      (requestedFields.children &&
+        Object.keys(requestedFields.children).some(key => key !== 'id'))
+    ) {
+      // count all parents and children in the returned spaces list, to make sure we don't exceed the max
+      // otherwise throw explanatory error
+      const subRequestsToMake = spaces.reduce((a, space) => {
+        if (space.children) a += space.children.length;
+        if (space.parent) a++;
+        return a;
+      }, 0);
 
-  let { first = 20 } = args;
-  const { skip = 0 } = args;
-  if (first > 1000) first = 1000;
-  params.push(skip, first);
+      if (subRequestsToMake > maxSubRequests) {
+        throw new Error(
+          `Too many subrequests to make. (Max: ${maxSubRequests}) Query only 'id' for parent and children nodes or reduce result limit.`
+        );
+      }
 
-  const query = `
-    SELECT s.* FROM spaces s
-    WHERE 1 = 1 ${queryStr}
-    GROUP BY s.id
-    ORDER BY s.${orderBy} ${orderDirection} LIMIT ?, ?
-  `;
-  try {
-    const spaces = await db.queryAsync(query, params);
-    return spaces.map(space =>
-      Object.assign(space, formatSpace(space.id, space.settings))
-    );
+      // fetch parents and children
+      for (const space of spaces) {
+        if (space.parent?.id) {
+          const parentResult = await fetchSpaces({
```

Comment:
Could you also do "in" query and get the list of spaces at once without needing to do multiple queries?

Reply:
If it works, then you wont have to calculate and error out when there are more than 100 children.

Reply:
Aaaahhh yea sure that should be a lot better. Collect all ids, make one query and then remap everything in js. Gonna do that.

Reply:
So much better. Not sure why I over-complicated it like that. I also moved the check whether parent and child match, into the resolver now. Less work in clients. Gonna update UI now accordingly.